When adding image support for "com.adobe.photoshop-image", also add MIME type "application/x-photoshop"

squelart opened the pull request:
#### b2fab9cbd1d1baf5a80ac164ed4bf46179aabb19
<pre>
When adding image support for &quot;com.adobe.photoshop-image&quot;, also add MIME type &quot;application/x-photoshop&quot;
<a href="https://bugs.webkit.org/show_bug.cgi?id=260464">https://bugs.webkit.org/show_bug.cgi?id=260464</a>
rdar://103614570

Reviewed by Said Abou-Hallawa.

Mail adds &quot;com.adobe.photoshop-image&quot; as supported type, so that Photoshop images may be viewed as images
in emails. At that time, the associated MIME type &quot;image/vnd.adobe.photoshop&quot; also gets added.
However when inserting an attachment with file extension &quot;psd&quot;, the system infers its MIME type as
&quot;application/x-photoshop&quot; instead, which is not supported by default, this prevents psd&apos;s from appearing
in the Mail editor.

So when adding &quot;com.adobe.photoshop-image&quot;, the MIME type &quot;application/x-photoshop&quot; gets added as well to
correctly show attached psd images.

* Source/WebCore/platform/graphics/cg/UTIRegistry.cpp:
(WebCore::setAdditionalSupportedImageTypes):
* Source/WebCore/platform/network/mac/UTIUtilities.h:

* Source/WebCore/platform/network/mac/UTIUtilities.mm:
(WebCore::RequiredMIMETypesFromUTI):
Note: This 1:M UTI-&gt;MIME mapping should be exceedingly rare, so a simple `if` test here is enough.

* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/cocoa/TestUTIRegistry.cpp: Added.
(TestWebKitAPI::TEST):
Relevant tests for this fix, but also a check to know if one day this fix is not needed anymore or needs
an update (e.g., if the platform starts to associate *.psd files with &quot;image/vnd.adobe.photoshop&quot; or
another MIME type).

* Tools/TestWebKitAPI/Tests/WebCore/cocoa/TestUTIUtilities.cpp: Added.
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/267153@main">https://commits.webkit.org/267153@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/60638f3c7224112cb164cd13419364cbff3255bb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/15762 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16080 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/16456 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17524 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14792 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/15947 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/19073 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16175 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/17292 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15952 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16422 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13419 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18277 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/13670 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14231 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/21131 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14673 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14396 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17651 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14991 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/12708 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14238 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3773 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18607 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14813 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/platform/graphics/cg/UTIRegistry.cpp:
```diff
@@ -107,14 +107,38 @@ MemoryCompactRobinHoodHashSet<String>& additionalSupportedImageTypes()
+
+        auto mapEntry = additionalImageTypeMimeTypeMap().find<ASCIICaseInsensitiveStringViewHashTranslator>(imageType);
+        if (mapEntry != additionalImageTypeMimeTypeMap().end())
+            MIMETypeRegistry::additionalSupportedImageMIMETypes().add(mapEntry->value);
```
This is changing the mapping from 1:1 (UTI -> MimeType) to 1:M relation. So I would suggest moving this new relationship to a new function in UTIUtilities.cpp named something like `RequiredMIMETypesFromUTI()`. It can return a `HashSet<String>`. And we can replace the call above `MIMETypeForImageType()` by a call to this new function.

shallawa commented on Source/WebCore/platform/graphics/cg/UTIRegistry.cpp:
```diff
@@ -107,14 +107,38 @@ MemoryCompactRobinHoodHashSet<String>& additionalSupportedImageTypes()
+    }();
+    return mimeTypesMap;
+}
+
```
I think it is a very rare scenario which requires supporting more than one MIMEType for a given UTI. And I would not expect that we will be adding entries to this HashMap at least in the near future. So I would suggest just adding this as a single if-statement:

```
if (equalLettersIgnoringASCIICase(imageType, "com.adobe.photoshop-image"_s)
    return "application/x-photoshop"_s;
```

shallawa commented on Source/WebCore/platform/graphics/cg/UTIRegistry.cpp:
```diff
@@ -112,9 +112,8 @@ void setAdditionalSupportedImageTypes(const Vector<String>& imageTypes)
-        if (!mimeType.isEmpty())
-            MIMETypeRegistry::additionalSupportedImageMIMETypes().add(mimeType);
+        auto mimeTypes = RequiredMIMETypesFromUTI(imageType);
+        MIMETypeRegistry::additionalSupportedImageMIMETypes().add(mimeTypes.begin(), mimeTypes.end());
```
Can we add an API test for this change?

squelart committed:
When adding image support for "com.adobe.photoshop-image", also add MIME type "application/x-photoshop"
https://bugs.webkit.org/show_bug.cgi?id=260464
rdar://103614570

Reviewed by Said Abou-Hallawa.

Mail adds "com.adobe.photoshop-image" as supported type, so that Photoshop images may be viewed as images
in emails. At that time, the associated MIME type "image/vnd.adobe.photoshop" also gets added.
However when inserting an attachment with file extension "psd", the system infers its MIME type as
"application/x-photoshop" instead, which is not supported by default, this prevents psd's from appearing
in the Mail editor.

So when adding "com.adobe.photoshop-image", the MIME type "application/x-photoshop" gets added as well to
correctly show attached psd images.

* Source/WebCore/platform/graphics/cg/UTIRegistry.cpp:
(WebCore::setAdditionalSupportedImageTypes):
* Source/WebCore/platform/network/mac/UTIUtilities.h:

* Source/WebCore/platform/network/mac/UTIUtilities.mm:
(WebCore::RequiredMIMETypesFromUTI):
Note: This 1:M UTI->MIME mapping should be exceedingly rare, so a simple `if` test here is enough.

* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/cocoa/TestUTIRegistry.cpp: Added.
(TestWebKitAPI::TEST):
Relevant tests for this fix, but also a check to know if one day this fix is not needed anymore or needs
an update (e.g., if the platform starts to associate *.psd files with "image/vnd.adobe.photoshop" or
another MIME type).

* Tools/TestWebKitAPI/Tests/WebCore/cocoa/TestUTIUtilities.cpp: Added.
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/267153@main
